src/library/scala/collection/immutable/NVector.scala

```diff
@@ -0,0 +1,2004 @@
+      else if(i2 < suffix2.length) copy(suffix2 = copyUpdate(suffix2, i2, i1, elem))
+      else                         copy(suffix1 = copyUpdate(suffix1, i1, elem))
+    } else if(index >= len12) {
+      val io = index - len12
+      copy(prefix3 = copyUpdate(prefix3, io >>> BITS2, (io >>> BITS) & MASK, io & MASK, elem))
+    } else if(index >= len1) {
+      val io = index - len1
+      copy(prefix2 = copyUpdate(prefix2, io >>> BITS, io & MASK, elem))
+    } else {
+      copy(prefix1 = copyUpdate(prefix1, index, elem))
+    }
+  }
+
+  override def appended[B >: A](elem: B): NVector[B] = {
+    if     (suffix1.length < WIDTH  ) copy(suffix1 = copyAppend(suffix1, elem.asInstanceOf[AnyRef]), length = length+1)
+    else if(suffix2.length < WIDTH-1) copy(suffix2 = copyAppend(suffix2, suffix1), suffix1 = wrap1(elem), length = length+1)
+    else if(suffix3.length < WIDTH-1) copy(suffix3 = copyAppend(suffix3, copyAppend(suffix2, suffix1)), suffix2 = empty2, suffix1 = wrap1(elem), length = length+1)
+    else if(data4.length   < WIDTH-2) copy(data4   = copyAppend(data4, copyAppend(suffix3, copyAppend(suffix2, suffix1))), suffix3 = empty3, suffix2 = empty2, suffix1 = wrap1(elem), length = length+1)
+    else new NVector5(prefix1, len1, prefix2, len12, prefix3, len123, data4, (WIDTH-2)*WIDTH3 + len123, empty5, wrap4(copyAppend(suffix3, copyAppend(suffix2, suffix1))), empty3, empty2, wrap1(elem), length+1)
+  }
+
+  override def prepended[B >: A](elem: B): NVector[B] = {
+    if     (len1         < WIDTH  ) copy(copyPrepend1(elem, prefix1), len1+1, len12 = len12+1, len123 = len123+1, length = length+1)
+    else if(len12        < WIDTH2 ) copy(wrap1(elem), 1, copyPrepend(prefix1, prefix2), len12+1, len123 = len123+1, length = length+1)
+    else if(len123       < WIDTH3 ) copy(wrap1(elem), 1, empty2, 1, copyPrepend(copyPrepend(prefix1, prefix2), prefix3), len123+1, length = length+1)
+    else if(data4.length < WIDTH-2) copy(wrap1(elem), 1, empty2, 1, empty3, 1, copyPrepend(copyPrepend(copyPrepend(prefix1, prefix2), prefix3), data4), length = length+1)
+    else new NVector5(wrap1(elem), 1, empty2, 1, empty3, 1, wrap4(copyPrepend(copyPrepend(prefix1, prefix2), prefix3)), len123+1, empty5, data4, suffix3, suffix2, suffix1, length+1)
+  }
+
+  override def foreach[U](f: A => U): Unit = {
+    foreachElem(prefix1, f)
+    foreachElem(prefix2, f)
+    foreachElem(prefix3, f)
+    foreachElem(data4, f)
+    foreachElem(suffix3, f)
+    foreachElem(suffix2, f)
+    foreachElem(suffix1, f)
+  }
+
+  override def map[B](f: A => B): NVector[B] =
+    copy(prefix1 = mapElems1(prefix1, f), prefix2 = mapElems(2, prefix2, f), prefix3 = mapElems(3, prefix3, f),
+      data4 = mapElems(4, data4, f),
+      suffix3 = mapElems(3, suffix3, f), suffix2 = mapElems(2, suffix2, f), suffix1 = mapElems1(suffix1, f))
+
+  override def last: A = suffix1(suffix1.length-1).asInstanceOf[A]
+
+  protected[this] def slice0(lo: Int, hi: Int): NVector[A] = {
```

Comment:
Consider trying to reduce the bytecode count of this method, it might make it inlineable in all methods which call it.

Reply:
It's the same issue as in `foreach`. You can easily abstract using` vectorSlices` but the hand-tailored implementations specific to one dimension are faster.

Reply:
I ran some benchmarks. The generic version is about 20% slower (but still between 3x and millions of times faster compared to the O(n) implementation in Vector).

Reply:
@szeiger Hmmm, I'm not sure I understand what you mean.

Reply:
Reducing the size of the method means abstracting over slices in a loop. And if you do that you can move the method up and make it final, too. But it's 20% slower than the current version. I managed to do some other improvements in the slicing implementation but it doesn't affect this. A separate `slice0` that adds all slices manually is still 20% faster than the generic version.

Reply:
@szeiger Ah, that clears it up, thanks!